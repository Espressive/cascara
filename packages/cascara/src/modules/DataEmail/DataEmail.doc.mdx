---
title: Data Email
propTable: DataEmail.js
type: 'module'
status: 'Draft'
public: false
---

A data module to display **email** values, it renders a input of type **email**.

## Using the module

Additional to the props specified aside, please use the following property to configure your module:

| property | type                | description                                          |
| -------- | ------------------- | ---------------------------------------------------- |
| module   | required **String** | specifies the type of module, in this case **email** |

Adding this module to your **Cascara** **Table** or **Form** would look like this:

```jsx
<Table
  // we need to enable edition by passing the `edit` module
  actions={{
    actionButtonMenuIndex: -1,
    modules: [
      {
        module: 'edit',
        cancelLabel: 'Cancelar',
        editLabel: 'Editar',
        saveLabel: 'Guardar',
      },
    ],
  }}
  // your data
  data={[
    {
      id: 1,
      email: 'test@example.com',
    },
  ]}
  // the fields in our data you want to display
  dataConfig={{
    display: [
      {
        attribute: 'email',
        isEditable: true,
        isLabeled: false,
        label: 'eMail',
        module: 'email',
      },
    ],
  }}
  // the action handler is invoken when a button is clicked
  onAction={(action, record) => {
    console.clear();
    console.log(`Action ${action.name} invoked`);
    console.table(record);

    switch (action.name) {
      case 'edit.start':
        console.log('Entering edit mode');
        break;
      case 'edit.cancel':
        console.log('Exiting edit mode, edit canceled');
        break;
      case 'edit.save':
        console.log('Exitin edit mode, saving changes');
        break;
    }
  }}
  // this defines the unique ID field in your data
  uniqueIdAttribute={'id'}
/>
```

> Open your console!
