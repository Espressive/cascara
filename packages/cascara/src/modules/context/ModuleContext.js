import { createContext } from 'react';

// We should ONLY put values here that are needed for ALL Modules
const defaultValue = {
  isEditing: false,
  isEditingAny: false,
  value: 'blue',
};

/*
 * We will use this baseline context to create any of our context providers from.
 * Any unique functionality should be included in that provider and not in this Context.
 */
const ModuleContext = createContext(defaultValue);

/*
 * We are also exporting the default value in case we need to hydrate or merge
 * any providers with these default values.
 */
export { defaultValue };

export default ModuleContext;
