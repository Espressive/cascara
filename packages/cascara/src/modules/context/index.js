export { default as ModuleContext } from './ModuleContext';
export { default as ModuleProvider } from './ModuleProvider';
