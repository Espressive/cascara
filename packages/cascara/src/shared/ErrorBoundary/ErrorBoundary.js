import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import DefaultErrorTemplate from './templates/DefaultErrorTemplate';

/**
 * A ErrorBoundary will catch and display an error
 */

class ErrorBoundary extends PureComponent {
  static propTypes = {
    children: PropTypes.element,
    errorTemplate: PropTypes.func,
  };

  static defaultProps = {
    children: null,
    errorTemplate: DefaultErrorTemplate,
  };

  state = {
    error: null,
    errorInfo: null,
  };

  // Catch errors in any components below and re-render with error message
  // You can also log error messages to an error reporting service here
  componentDidCatch = (error, errorInfo) => {
    this.setState({
      error: error,
      errorInfo: errorInfo,
    });
  };

  render() {
    const { children, errorTemplate } = this.props;

    const { error, errorInfo } = this.state;

    if (errorInfo) {
      // Error path
      return React.createElement(errorTemplate, {
        error,
        errorInfo,
      });
    }
    // Normally, just render children
    return children;
  }
}

export default ErrorBoundary;
