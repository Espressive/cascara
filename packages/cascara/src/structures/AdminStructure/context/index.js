export { default as AdminContext } from './AdminContext';
export { default as AdminProvider } from './AdminProvider';
