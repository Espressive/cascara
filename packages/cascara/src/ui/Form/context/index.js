import FormProvider from './FormProvider';

export { FormProvider };
