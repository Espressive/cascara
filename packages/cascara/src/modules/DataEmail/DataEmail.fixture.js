import React from 'react';
import DataEmail from './DataEmail';
import ModuleSandbox from '../ModuleSandbox';

const DisplayDataEmail = ({ isEditing, value, ...rest }) => (
  <ModuleSandbox isEditing={isEditing}>
    <DataEmail {...rest} label='Email address' value={value} />
  </ModuleSandbox>
);

const emailValue = 'me@email.com';
const DataEmailModule = () => {
  return (
    <>
      <div>
        <h1>Email input field</h1>
        <p>
          This fixture is for testing the presentation of an input email field
        </p>
      </div>

      <hr />

      <div>
        <h2>Display</h2>
        <DisplayDataEmail value={emailValue} />
      </div>

      <div>
        <h2>Editing</h2>
        <DisplayDataEmail isEditing value={emailValue} />
      </div>
    </>
  );
};

export { DisplayDataEmail };
export default <DataEmailModule />;
