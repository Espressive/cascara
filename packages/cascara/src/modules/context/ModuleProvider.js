import React from 'react';
import pt from 'prop-types';

import ModuleContext, { defaultValue } from './ModuleContext';
import useToggle from '../../hooks/useToggle';

const propTypes = {
  children: pt.oneOfType([pt.element, pt.arrayOf(pt.element)]),
  value: pt.shape({
    isEditing: pt.bool,
  }),
};

const ModuleProvider = ({ children, value, ...props }) => {
  const [isEditing, setIsEditing] = useToggle(value?.isEditing || false);

  const mergedValues = {
    ...defaultValue,
    isEditing,
    setIsEditing,
    ...value,
  };

  return (
    <ModuleContext.Provider value={mergedValues} {...props}>
      {children}
    </ModuleContext.Provider>
  );
};

ModuleProvider.propTypes = propTypes;

export default ModuleProvider;
