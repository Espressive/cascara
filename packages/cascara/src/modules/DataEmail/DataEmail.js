import React, { useContext } from 'react';
import { Input } from 'reakit/Input';
import pt from 'prop-types';
import { ModuleContext } from '../context';
import styles from '../DataModule.module.scss';

import ErrorBoundary from '../../shared/ErrorBoundary';

const propTypes = {
  /** A Module can be defined to not present an editing state */
  isEditable: pt.bool,
  /** Presents the input without a label. NOT USER CONFIGURABLE */
  isLabeled: pt.bool,
  /** A Module needs to have a unique label relative to its context */
  label: pt.string.isRequired,
  /** A Module can have a value */
  value: pt.string,
};

const DataEmail = ({
  isEditable = true,
  isLabeled = true,
  label,
  value,
  ...rest
}) => {
  const { isEditing, formMethods } = useContext(ModuleContext);
  const renderEditing = (
    <label htmlFor={label}>
      {label && isLabeled && <span className={styles.Label}>{label}</span>}
      <Input
        {...rest}
        className={styles.Input}
        defaultValue={value}
        id={label}
        name={label}
        ref={formMethods?.register}
        type='email'
      />
    </label>
  );

  const renderDisplay = (
    <span>
      {label && isLabeled && <span className={styles.Label}>{label}</span>}
      <span className={styles.Input}>{value}</span>
    </span>
  );

  // Do not render an editable input if the module is not editable
  return (
    <ErrorBoundary>
      <div className={styles.Email}>
        {isEditing && isEditable ? renderEditing : renderDisplay}
      </div>
    </ErrorBoundary>
  );
};

DataEmail.propTypes = propTypes;

export default DataEmail;
