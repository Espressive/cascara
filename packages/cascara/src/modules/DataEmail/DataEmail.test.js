import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import DataEmail from './DataEmail';
import { DisplayDataEmail } from './DataEmail.fixture';

describe('Input Date', () => {
  // without ModuleSandbox will render the property information into a span
  describe('default', () => {
    const testId = 'default';

    // We need a place to store the view for snapshot testing. This is not required when we are using `screen` directly from RTL.
    let view;

    beforeEach(() => {
      // Set the render container to our `view` so it is in scope for the snapshot test
      view = render(<DataEmail data-testid={testId} />).container;
    });

    test('renders without any props', () => {
      expect(view).toMatchSnapshot();
    });

    test('renders a <span> by default', () => {
      const input = screen.getByTestId(testId);
      // Make sure the actual DOM element is not render an input
      expect(input.tagName).toMatch('SPAN');
    });
  });

  describe('editing', () => {
    // specifying ModuleSandbox will have the flag to render our input date field
    const testId = 'default';
    beforeEach(() =>
      render(<DisplayDataEmail data-testid={testId} isEditing />)
    );
    test('renders a <input email> by default', () => {
      const input = screen.getByTestId(testId);
      // Check that we also use the correct type for accessibility
      expect(input).toHaveAttribute('type', 'email');
    });

    test('change value', () => {
      const newEmail = 'newme@mail.com';
      const input = screen.getByTestId(testId);
      userEvent.type(input, newEmail);
      expect(input).toHaveValue(newEmail);
    });
  });
});
