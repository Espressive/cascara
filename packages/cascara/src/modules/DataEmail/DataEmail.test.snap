// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Date default renders without any props 1`] = `
<div>
  <div
    class="Email"
  >
    <span>
      <span
        class="Input"
        data-testid="default"
      />
    </span>
  </div>
</div>
`;
